### ISSUE TITLE
Async

### ISSUE THREAD

**Issue by brentvee** [NakedObjectsGroup/NakedObjectsFramework#85]
Hey Richard,
Hope you're well.
I'm investigating the newly released NOF 9.
I would like to know if there is support or plans for using ```async``` ?
Specifically for ```Container.SaveChanges``` and then for enumerating for example  
```var people = await Container.Instances<Person>().ToListAsync();```

**Comment by richardpawson** [comment]
Brent

You can certainly use async techniques within domain code that runs under naked objects, but we don't have any plans to provide explicit async support at the container level, and I am not sure I could see the need for it.  The most common need for async behaviour is in the client, where we are now running entirely under Angular, which has its own forms of asynchronous behaviour.

Richard

**Comment by brentvee** [comment]
Let me try explaining the motivation for my question a bit more.
Using the C# async and await keywords allows us to consume less thread resources while the system is busy with I/O operations.
For example, the following line that loads a person entity from the DB, will consume the thread while the I/O work of reading the record from the DB and sending it over the network is busy happening.
```var p = Container.Instances<Person>().Single(p => p.Id == 1);```

However, the following version will release the thread while the I/O work is busy, making the thread available to do other work.
```var p = await Container.Instances<Person>().SingleAsync(p => p.Id == 1);```

The same goes for ```SaveChanges()``` vs. ```await SaveChangesAsync()```

In other words, when the code starts handling many requests, the async versions allow much more efficient utilisation of threads and, as a result CPU cores.
The non async version consumes much more CPU, and requires horizontal scaling much sooner.

We currently have to scale out to 6 to 8 servers with 4 cores each during our business times because of this.

So we want to know if you have plans to support async and await in the framework during a HTTP request’s life cycle.

E.g. if one of my methods returns an ```IQueryable<Person>``` that NOF renders, it should not call ```q.Take(20).ToList()``` to materialize it from the DB , but rather ```await q. .Take(20).ToListAsync()```

Does that make sense?

**Comment by richardpawson** [comment]
I understand the principle, I am just a bit dubious about the likely net impact on your performance -  but am happy to be shown to be wrong.

In any event, it is still not clear to me that this requires any change to NOF.  I have not personally ever tried these ...Async LINQ methods on a queryable.  Do they just work as is?  All that Container.Instances<T> does is delegate (using a reflection trick) to the appropriate property on the DbContext, returning the appropriate DbSet<T> from that property.  So if you can do your async stuff on standard Entity Framework then it should work in Naked Objects  -  and if you can't then this is not an issue that we can address.

**Comment by brentvee** [comment]
So I tried a basic implementation and NOF does not render this service action.
I'm guessing its because the return type is a Task and not a domain entity
```
public async Task<Person> ListPersonsAsync(Guid id)
{
	var count = await Container.Instances<Person>().CountAsync();
	var person = await Container.Instances<Person>().SingleOrDefaultAsync(o => o.Id == id);
	return person;
}
```

**Comment by richardpawson** [comment]
Yes, you're obviously going to have to let the async task complete _before_ you return the result  -  because the result has to be sent over the RESTful API.

**Comment by brentvee** [comment]
Hey Richard,
We're still using NOF 7.
So I'm guessing this isn't going to be possible?

**Comment by richardpawson** [comment]
Your post started by saying you were exploring NOF9.  But it doesn't make much difference. Any async processes within domain code will need to have completed before you attempt to return any results dependent upon those async processes to the user.  And, I have to say, it would make no sense to me to attempt to attempt to make it do otherwise.  Given that the NOF server is stateless, every user request is going to require some objects to be retrieved from the database (or, perhaps, cache) -  and that tends to be the throttle on performance. So freeing up a server thread while you were waiting for a database process to complete isn't actually going to give you any net additional capacity, is it (unless your database server can handle a lot more simultaneous requests than your app server can)?
